### PR TITLE
LTD-5154: Add dedicated endpoint to archive/restore a good

### PR DESCRIPTION
**Pull request by saruniitr**
## Change description

Exporter has a choice to archive/restore a good and this is only allowed for goods that are submitted/verified on an application.

However we have a restriction that goods cannot be edited once they are submitted on an application. Removing this restriction can cause various issues so to avoid this add a dedicated endpoint which only allows archiving/restoring a particular good.

Add tests to ensure that only archive status can be updated using this endpoint.

### CODE REVIEW

**Review by seijihg — APPROVED** [review]

File: api/goods/tests/test_edit.py [hunk 1]
```diff
@@ -593,6 +593,37 @@ def test_edit_archive_status(self):
...
+                {"name": "Rifle", "is_archived": False},
+            ],
+            [
+                {"name": "Rifle", "is_archived": None},
+                {"name": "Rifle updated", "is_archived": False},
+                {"name": "Rifle", "is_archived": False},
+            ],
+        ]
+    )
+    def test_only_archive_field_can_be_updated(self, initial, data, expected):
```
Comment (currycoder) [inline api/goods/tests/test_edit.py:615]: Was about to comment recommending something like this, but here it is!

File: api/goods/views.py [hunk 2]
```diff
@@ -194,6 +195,12 @@ def get_queryset(self):
         return queryset.order_by("-updated_at")
 
 
+class GoodArchiveRestore(UpdateAPIView):
+    authentication_classes = (ExporterAuthentication,)
+    queryset = Good.objects.all()
```
Comment (currycoder) [inline api/goods/views.py:200]: Could I suggest a custom queryset method which filters the eligible Goods down to those that are associated with the user's organisation.  e.g. a double defence against somebody being able to archive another org's goods in some way

Comment (saruniitr) [reply api/goods/views.py:200]: yeah good idea, will further limit to 'submitted, verified' goods

File: api/goods/views.py [hunk 3]
```diff
@@ -194,6 +195,19 @@ def get_queryset(self):
...
 
 
+class GoodArchiveRestore(UpdateAPIView):
+    authentication_classes = (ExporterAuthentication,)
+    serializer_class = GoodArchiveRestoreSerializer
+
+    def get_queryset(self):
+        organisation = get_request_user_organisation_id(self.request)
+
+        return Good.objects.filter(
```
Comment (depsiatwal) [inline api/goods/views.py:205]: nicely done

File: api/goods/tests/test_edit.py [hunk 4]
```diff
@@ -593,6 +598,70 @@ def test_edit_archive_status(self):
...
+                {"name": "Rifle", "is_archived": True},
+                {"is_archived": False},
+                {"name": "Rifle", "is_archived": False},
+            ],
+            [
+                {"name": "Rifle", "is_archived": None},
+                {"name": "Rifle updated", "is_archived": False},
+                {"name": "Rifle", "is_archived": False},
+            ],
+        ]
```
Comment (depsiatwal) [inline api/goods/tests/test_edit.py:618]: Are we missing {"name": "Rifle", "is_archived": False}, ?

File: api/goods/views.py [hunk 5]
```diff
@@ -194,6 +195,19 @@ def get_queryset(self):
...
+            organisation=organisation,
```
Comment (kevincarrogan) [inline api/goods/views.py:206]: This looks like a fairly common pattern to filter by organisation. I wonder if extracting this out as a DRF FilterBackend would be useful.

File: api/goods/tests/test_edit.py (hunk 4 above)
Comment (saruniitr) [reply api/goods/tests/test_edit.py:618]: Actually for this test it doesn't really matter.
`is_archived = False` is for restoring case.

The main intention of this test is to ensure only `is_archived` is updated to either `True/False` and other fields remain unmodified.

**Review by depsiatwal — APPROVED** [review]
lgtm

File: api/goods/views.py (hunk 5 above)
Comment (saruniitr) [reply api/goods/views.py:206]: organisation id is not part of the url so more changes required, ok to take this up as a follow up?

**Pull request merged**